File: agent/internal/protocol/protocol.go
// Package protocol mirrors hub/app/ingest/protocol.py (proto v1).
// Changes within a version are additive only; unknown JSON fields are ignored
// by both sides.
package protocol

import "encoding/json"

const Version = 1

// ── agent → hub ─────────────────────────────────────────────────────────────

type Hello struct {
	T            string   `json:"t"`
	Proto        int      `json:"proto"`
	AgentVersion string   `json:"agent_version"`
	Hostname     string   `json:"hostname"`
	OS           string   `json:"os"`
	Arch         string   `json:"arch"`
	BootTS       int64    `json:"boot_ts"`
	Caps         []string `json:"caps"`
	// Current seq counter position; 0 on a fresh process so the hub resets
	// its dedup position instead of dropping every new frame.
	StartSeq uint64 `json:"start_seq"`
	// Declarative node UI; zero values mean "not declared".
	UIPort int    `json:"ui_port,omitempty"`
	UIUrl  string `json:"ui_url,omitempty"`
}

type Sample struct {
	Name  string  `json:"name"`
	Value float64 `json:"value"`
}

type Metrics struct {
	T       string   `json:"t"`
	Seq     uint64   `json:"seq"`
	TS      int64    `json:"ts"`
	Samples []Sample `json:"samples"`
}

type Heartbeat struct {
	T   string `json:"t"`
	Seq uint64 `json:"seq"`
	TS  int64  `json:"ts"`
}

// FsMountInfo mirrors the hub's FsMountInfo schema.
type FsMountInfo struct {
	Mountpoint string `json:"mountpoint"`
	Device     string `json:"device"`
	Fstype     string `json:"fstype"`
	TotalBytes int64  `json:"total_bytes"`
	UsedBytes  int64  `json:"used_bytes"`
	Stale      bool   `json:"stale"`
}

type Fs struct {
	T      string        `json:"t"`
	Seq    uint64        `json:"seq"`
	TS     int64         `json:"ts"`
	Mounts []FsMountInfo `json:"mounts"`
}

// ContainerInfo mirrors the hub's ContainerInfo schema.
type ContainerInfo struct {
	ContainerID string            `json:"container_id"`
	Name        string            `json:"name"`
	Image       string            `json:"image"`
	State       string            `json:"state"`
	Health      string            `json:"health"`
	Ports       []string          `json:"ports"`
	Labels      map[string]string `json:"labels"`
	StartedAt   int64             `json:"started_at"`
	// "host" containers list their EXPOSE'd ports in Ports — they listen on
	// the host directly. Additive field; the hub defaults it to empty.
	NetworkMode string `json:"network_mode,omitempty"`
}

type ContainersFull struct {
	T          string          `json:"t"`
	Seq        uint64          `json:"seq"`
	TS         int64           `json:"ts"`
	Containers []ContainerInfo `json:"containers"`
}

type ContainerEvent struct {
	T         string        `json:"t"`
	Seq       uint64        `json:"seq"`
	TS        int64         `json:"ts"`
	Action    string        `json:"action"`
	Container ContainerInfo `json:"container"`
}

// ContainerStat mirrors the hub's ContainerStat schema. CPUPct is nil on the
// first sample: usage percentages need a delta between two reads.
type ContainerStat struct {
	ContainerID string   `json:"container_id"`
	CPUPct      *float64 `json:"cpu_pct"`
	MemBytes    *uint64  `json:"mem_bytes"`
	MemPct      *float64 `json:"mem_pct"`
}

type ContainerStats struct {
	T     string          `json:"t"`
	Seq   uint64          `json:"seq"`
	TS    int64           `json:"ts"`
	Stats []ContainerStat `json:"stats"`
}

// SmartDisk mirrors the hub's SmartDisk schema.
type SmartDisk struct {
	Device         string   `json:"device"`
	Model          string   `json:"model"`
	Serial         string   `json:"serial"`
	Kind           string   `json:"kind"`
	CapacityBytes  int64    `json:"capacity_bytes"`
	SmartStatus    string   `json:"smart_status"`
	TempC          *float64 `json:"temp_c"`
	PowerOnHours   *int64   `json:"power_on_hours"`
	ReallocSectors *int64   `json:"realloc_sectors"`
	PendingSectors *int64   `json:"pending_sectors"`
	WearPct        *float64 `json:"wear_pct"`
	RawJSON        string   `json:"raw_json"`
}

type Smart struct {
	T     string      `json:"t"`
	Seq   uint64      `json:"seq"`
	TS    int64       `json:"ts"`
	Disks []SmartDisk `json:"disks"`
}

type K8sDetected struct {
	T           string `json:"t"`
	Seq         uint64 `json:"seq"`
	TS          int64  `json:"ts"`
	Distro      string `json:"distro"`
	Version     string `json:"version"`
	APIEndpoint string `json:"api_endpoint"`
	Kubeconfig  string `json:"kubeconfig"`
}

// SpeedtestResult answers a hub-triggered speedtest request.
type SpeedtestResult struct {
	T            string  `json:"t"`
	Seq          uint64  `json:"seq"`
	TS           int64   `json:"ts"`
	RequestID    int64   `json:"request_id"`
	LatencyMs    float64 `json:"latency_ms"`
	DownloadMbps float64 `json:"download_mbps"`
	UploadMbps   float64 `json:"upload_mbps"`
	Error        string  `json:"error,omitempty"`
}

// Sequenced is any agent→hub data frame the transport numbers and buffers.
type Sequenced interface{ SetSeq(seq uint64) }

func (m *Metrics) SetSeq(seq uint64)         { m.Seq = seq }
func (m *Fs) SetSeq(seq uint64)              { m.Seq = seq }
func (m *ContainersFull) SetSeq(seq uint64)  { m.Seq = seq }
func (m *ContainerEvent) SetSeq(seq uint64)  { m.Seq = seq }
func (m *ContainerStats) SetSeq(seq uint64)  { m.Seq = seq }
func (m *Smart) SetSeq(seq uint64)           { m.Seq = seq }
func (m *K8sDetected) SetSeq(seq uint64)     { m.Seq = seq }
func (m *SpeedtestResult) SetSeq(seq uint64) { m.Seq = seq }

func NewHello(agentVersion, hostname, osName, arch string, bootTS int64, caps []string) Hello {
	return Hello{
		T: "hello", Proto: Version, AgentVersion: agentVersion,
		Hostname: hostname, OS: osName, Arch: arch, BootTS: bootTS, Caps: caps,
	}
}

func NewMetrics(ts int64, samples []Sample) *Metrics {
	return &Metrics{T: "metrics", TS: ts, Samples: samples}
}

func NewFs(ts int64, mounts []FsMountInfo) *Fs {
	if mounts == nil {
		mounts = []FsMountInfo{}
	}
	return &Fs{T: "fs", TS: ts, Mounts: mounts}
}

func NewContainersFull(ts int64, containers []ContainerInfo) *ContainersFull {
	return &ContainersFull{T: "containers_full", TS: ts, Containers: containers}
}

func NewContainerStats(ts int64, stats []ContainerStat) *ContainerStats {
	return &ContainerStats{T: "container_stats", TS: ts, Stats: stats}
}

func NewContainerEvent(ts int64, action string, container ContainerInfo) *ContainerEvent {
	// Normalize nil slices/maps: they marshal as JSON null otherwise.
	if container.Ports == nil {
		container.Ports = []string{}
	}
	if container.Labels == nil {
		container.Labels = map[string]string{}
	}
	return &ContainerEvent{T: "container_event", TS: ts, Action: action, Container: container}
}

func NewSmart(ts int64, disks []SmartDisk) *Smart {
	if disks == nil {
		disks = []SmartDisk{}
	}
	return &Smart{T: "smart", TS: ts, Disks: disks}
}

func NewK8sDetected(ts int64, distro, apiEndpoint, kubeconfig string) *K8sDetected {
	return &K8sDetected{
		T: "k8s_detected", TS: ts, Distro: distro,
		APIEndpoint: apiEndpoint, Kubeconfig: kubeconfig,
	}
}

func NewHeartbeat(seq uint64, ts int64) Heartbeat {
	return Heartbeat{T: "heartbeat", Seq: seq, TS: ts}
}

// ── hub → agent ─────────────────────────────────────────────────────────────

type AgentConfig struct {
	MetricsIntervalS   int `json:"metrics_interval_s"`
	FsIntervalS        int `json:"fs_interval_s"`
	SmartIntervalS     int `json:"smart_interval_s"`
	HeartbeatIntervalS int `json:"heartbeat_interval_s"`
}

type HelloAck struct {
	T             string      `json:"t"`
	Proto         int         `json:"proto"`
	NodeUUID      string      `json:"node_uuid"`
	AgentToken    string      `json:"agent_token"`
	Config        AgentConfig `json:"config"`
	ResumeFromSeq uint64      `json:"resume_from_seq"`
}

type Ack struct {
	T       string `json:"t"`
	UptoSeq uint64 `json:"upto_seq"`
}

type ConfigUpdate struct {
	T      string      `json:"t"`
	Config AgentConfig `json:"config"`
}

type ErrorMsg struct {
	T    string `json:"t"`
	Code string `json:"code"`
	Msg  string `json:"msg"`
}

// Envelope peeks the discriminator so the reader can decode the right type.
type Envelope struct {
	T string `json:"t"`
}

func MessageType(raw []byte) (string, error) {
	var env Envelope
	if err := json.Unmarshal(raw, &env); err != nil {
		return "", err
	}
	return env.T, nil
}

File: hub/app/ingest/protocol.py
"""Agent ↔ hub protocol v1.

This module is the canonical contract; `agent/internal/protocol/` mirrors it in Go.
Within a proto version changes must be additive only — both sides ignore unknown
fields. Bump PROTO_VERSION on breaking changes; the hub accepts N and N-1.
"""

from typing import Annotated, Literal

from pydantic import BaseModel, BeforeValidator, ConfigDict, Field, TypeAdapter

PROTO_VERSION = 1
MIN_PROTO_VERSION = 1


def _none_to_list(v: object) -> object:
    return [] if v is None else v


def _none_to_dict(v: object) -> object:
    return {} if v is None else v


# Go marshals nil slices/maps as JSON null; tolerate that everywhere.
StrList = Annotated[list[str], BeforeValidator(_none_to_list)]
StrMap = Annotated[dict[str, str], BeforeValidator(_none_to_dict)]


class _Msg(BaseModel):
    model_config = ConfigDict(extra="ignore")


# ── agent → hub ──────────────────────────────────────────────────────────────


class Hello(_Msg):
    t: Literal["hello"]
    proto: int
    agent_version: str = ""
    hostname: str = ""
    os: str = ""
    arch: str = ""
    boot_ts: int = 0
    caps: StrList = []
    # Where the agent's seq counter currently stands. A freshly restarted
    # (stateless) agent declares 0 so the hub resets its dedup position
    # instead of discarding every new frame as a duplicate.
    start_seq: int = 0


class Sample(_Msg):
    name: str
    value: float


class Metrics(_Msg):
    t: Literal["metrics"]
    seq: int
    ts: int
    samples: list[Sample] = []


class FsMountInfo(_Msg):
    mountpoint: str
    device: str = ""
    fstype: str = ""
    total_bytes: int = 0
    used_bytes: int = 0
    stale: bool = False


class Fs(_Msg):
    t: Literal["fs"]
    seq: int
    ts: int
    mounts: list[FsMountInfo]


class ContainerInfo(_Msg):
    container_id: str
    name: str
    image: str = ""
    state: str = ""
    health: str = ""
    ports: StrList = []
    labels: StrMap = {}
    started_at: int = 0
    # "host" containers list their EXPOSE'd ports in `ports` (they listen on
    # the host directly); additive field, absent from older agents.
    network_mode: str = ""


class ContainersFull(_Msg):
    t: Literal["containers_full"]
    seq: int
    ts: int
    containers: list[ContainerInfo]


class ContainerEvent(_Msg):
    t: Literal["container_event"]
    seq: int
    ts: int
    action: str  # start | die | health_status | ...
    container: ContainerInfo


class ContainerStat(_Msg):
    container_id: str
    cpu_pct: float | None = None  # None on the first sample (needs a delta)
    mem_bytes: int | None = None
    mem_pct: float | None = None


class ContainerStats(_Msg):
    t: Literal["container_stats"]
    seq: int
    ts: int
    stats: list[ContainerStat] = []


class SmartDisk(_Msg):
    device: str
    model: str = ""
    serial: str = ""
    kind: str = ""  # hdd | ssd | nvme
    capacity_bytes: int = 0
    smart_status: str = "unknown"  # passed | failed | unknown
    temp_c: float | None = None
    power_on_hours: int | None = None
    realloc_sectors: int | None = None
    pending_sectors: int | None = None
    wear_pct: float | None = None
    raw_json: str = ""


class Smart(_Msg):
    t: Literal["smart"]
    seq: int
    ts: int
    disks: list[SmartDisk]


class K8sDetected(_Msg):
    t: Literal["k8s_detected"]
    seq: int
    ts: int
    distro: str = ""  # k3s | kubeadm | k0s
    version: str = ""
    api_endpoint: str = ""
    kubeconfig: str = ""  # content, when readable


class Heartbeat(_Msg):
    t: Literal["heartbeat"]
    seq: int
    ts: int


class Speedtest(BaseModel):
    """Hub→agent: run a speedtest and answer with the same id."""

    t: Literal["speedtest"] = "speedtest"
    id: int


class SpeedtestResult(_Msg):
    t: Literal["speedtest_result"]
    seq: int
    ts: int
    request_id: int = 0
    latency_ms: float | None = None
    download_mbps: float | None = None
    upload_mbps: float | None = None
    error: str = ""


AgentToHub = Annotated[
    Hello
    | Metrics
    | Fs
    | ContainersFull
    | ContainerEvent
    | ContainerStats
    | Smart
    | K8sDetected
    | SpeedtestResult
    | Heartbeat,
    Field(discriminator="t"),
]

_agent_adapter: TypeAdapter[AgentToHub] = TypeAdapter(AgentToHub)


def parse_agent_message(raw: str | bytes) -> AgentToHub:
    return _agent_adapter.validate_json(raw)


# ── hub → agent ──────────────────────────────────────────────────────────────


class AgentConfig(_Msg):
    metrics_interval_s: int = 10
    fs_interval_s: int = 60
    smart_interval_s: int = 1800
    heartbeat_interval_s: int = 15


class HelloAck(_Msg):
    t: Literal["hello_ack"] = "hello_ack"
    proto: int = PROTO_VERSION
    node_uuid: str
    agent_token: str | None = None  # only on enrollment
    config: AgentConfig = AgentConfig()
    resume_from_seq: int = 0


class Ack(_Msg):
    t: Literal["ack"] = "ack"
    upto_seq: int


class ConfigUpdate(_Msg):
    t: Literal["config"] = "config"
    config: AgentConfig


class Resync(_Msg):
    t: Literal["resync"] = "resync"


class ErrorMsg(_Msg):
    t: Literal["error"] = "error"
    code: str
    msg: str = ""


class Bye(_Msg):
    t: Literal["bye"] = "bye"
